Some updates to gas and delegatecall. (#12910)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

Gas
===

Upon creation, each transaction is charged with a certain amount of **gas**
that has to be paid for by the originator of the transaction (``tx.origin``).
While the EVM executes the
transaction, the gas is gradually depleted according to specific rules.
If the gas is used up at any point (i.e. it would be negative),
an out-of-gas exception is triggered, which ends execution and reverts all modifications

> made to the state in the current call frame. 

Reviewer: I wanted to comment that it's not just the current call frame but the whole transaction since you cannot silence out-of-gas. Then I realized the caller could forward less gas and handle the revert. Maybe it would be a good idea to mention that the caller can decide how much gas he forwards?

> made to the state in the current call frame. 

Reviewer: Looks like `chk_coding_style` failed because of trailing whitespace here:
```suggestion
made to the state in the current call frame.
```

There exists a special variant of a message call, named **delegatecall**
which is identical to a message call apart from the fact that
the code at the target address is executed in the context (i.e. at the address) of the calling

> contract and ``msg.sender`` and ``msg.value`` do not change thair values.

Reviewer: ```suggestion
contract and ``msg.sender`` and ``msg.value`` do not change their values.
```

Reviewer 2: I must be going blind...

Reviewer: Fixed.

> contract and ``msg.sender`` and ``msg.value`` do not change thair values.

Reviewer: Since this is an intro, might be worth adding a few words to clarify the consequences:
```suggestion
contract and ``msg.sender`` and ``msg.value`` retain the values they have in that context,
effectively letting the callee impersonate the calling contract and freely modify its state.
```

Reviewer: @ekpyron What do you think about this one? Once we resolve this, I could fix the two typos, merge and we'd be one PR lighter :)

Reviewer 2: Hm... "context" is a bit underdefined for that :-). Earlier we say by "context" we mean "at the address of the calling
contract"... but `msg.sender` and `msg.value` don't have values "at that address"... they retain the values they had at the call site...

And in general the sentence is getting too long... I'd rather keep the sentence as in the PR now - and we can think about adding a new sentence about "effectively allowing to impersonate"... on the other hand the next paragraph also kind of says that...

Reviewer: > on the other hand the next paragraph also kind of says that...

A, right. Didn't look too far ahead :) OK, let's drop this suggestion in that case.

> This means that a contract can dynamically load code from a different
> address at runtime. Storage, current address and balance still

Reviewer 2: ```suggestion
This means that a contract can dynamically load code from a different
address at runtime and execute that code as if it was one of its internal functions. Storage, current address and balance still
```

Maybe that's a good way to clarify what you wanted to have in https://github.com/ethereum/solidity/pull/12910/files#r845484979 ? Executing the foreign code as if it was an internal function is basically the opposite way of putting the called code impersonating the calling contract, but to me it would seem easier to understand what it means...

Reviewer: Sounds good to me. I'll add that to the PR then.

Reviewer 2: On the other hand "as if it was one of its internal functions" might read like the called contract would somehow magically get access to the symbols in the calling contracts scope... but maybe I'm just slowly going mad over thinking about docs ;-).

Reviewer: Maybe. I thought it still gives you a better picture of what will happen.

But yeah, could be misleading. Let's just merge it as is with only typos fixes.

Reviewer 2: Yeah, I'm not sure anymore, maybe I'm starting to overthink what stuff might imply or not ;-). I'm fine with adding this and with not adding this.
